### PR TITLE
OBPIH-7879 Fix AvailableItem statuses logic  

### PR DESCRIPTION
**Pull request by SebastianLib**
### :sparkles: Description of Change

[//]: <> (A concise summary of what is being changed. Please provide enough context for reviewers to be able to understand the change and why it is necessary.)

**Link to GitHub issue or Jira ticket:**
https://pihemr.atlassian.net/browse/OBPIH-7830

**Description:**
Please read my comment before reviewing: https://pihemr.atlassian.net/browse/OBPIH-7830?focusedCommentId=193420

---
### :camera: Screenshots & Recordings (optional)

[//]: <> (If this PR contains a UI change, consider attaching one or more screenshots or recordings to help reviewers visualize the change. Otherwise, you can remove this section.)

### CODE REVIEW

File: grails-app/domain/org/pih/warehouse/product/Product.groovy [hunk 1]
```diff
@@ -495,12 +495,12 @@ class Product implements Comparable, Serializable, Validatable<ProductValidator>
...
-        // we first try to find inventory item with null lot number.
+        // we first try to find inventory items with null lot number, preferring the one without expiry.
         // TODO: After cleaning up the data we can remove this logic and always use the lot number that
         // TODO: we decided for default inventory items.
-        InventoryItem inventoryItemWithNullLot = InventoryItem.findByProductAndLotNumberIsNull(this)
-        if (inventoryItemWithNullLot) {
-            return inventoryItemWithNullLot
+        List<InventoryItem> inventoryItemsWithNullLot = InventoryItem.findAllByProductAndLotNumberIsNull(this)
+        if (inventoryItemsWithNullLot) {
+            return inventoryItemsWithNullLot.find { it.expirationDate == null } ?: inventoryItemsWithNullLot.get(0)
```
Comment (kchelstowski) [inline grails-app/domain/org/pih/warehouse/product/Product.groovy]: what if all of them would have expirationDate? or was it supposed not to be achievable? I don't remember what Kuba said

File: grails-app/services/org/pih/warehouse/inventory/InventoryService.groovy [hunk 2]
```diff
@@ -1705,7 +1703,7 @@ class InventoryService implements ApplicationContextAware {
             }
         }
         if (inventoryItems) {
-            return inventoryItems.get(0)
+            return inventoryItems.find { it.expirationDate == null } ?: inventoryItems.get(0)
```
Comment (kchelstowski) [inline grails-app/services/org/pih/warehouse/inventory/InventoryService.groovy]: same question as above

File: grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy [hunk 3]
```diff
@@ -1501,8 +1501,8 @@ class ShipmentService {
 
     def addTransactionEntries(Transaction transaction, Shipment shipment) {
         shipment.shipmentItems.each { ShipmentItem shipmentItem ->
-            def inventoryItem =
-                    inventoryService.findInventoryItemByProductAndLotNumber(shipmentItem.product, shipmentItem.lotNumber)
+            def inventoryItem = shipmentItem.inventoryItem
+                    ?: inventoryService.findInventoryItemByProductAndLotNumber(shipmentItem.product, shipmentItem.lotNumber)
```
Comment (kchelstowski) [inline grails-app/services/org/pih/warehouse/shipping/ShipmentService.groovy]: feels like a safe change, but I'm wondering why it wasn't like that before. 
If you touched this line, please add also the type, instead of `def`.

File: src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy [hunk 4]
```diff
@@ -25,6 +25,13 @@ class InventoryItemManager {
                                            Date expirationDate,
                                            boolean disableRefresh = false) {
         InventoryItem inventoryItem = getInventoryItem(product, lotNumber)
+
+        // If adding a default lot item (no lot, no expiry) but the existing inventoryItem has an expiration date,
+        // create a new one without expiry to avoid reusing it.
+        if (inventoryItem && !lotNumber && !expirationDate && inventoryItem.expirationDate) {
```
Comment (kchelstowski) [inline src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: could we shorten the conditions a bit, so:
```groovy
if (!inventoryItem || inventoryItem.expirationDate && !lotNumber && !expirationDate)) {
  return createInventoryItem(product, lotNumber, expirationDate, disableRefresh)
}

if (inventoryItem) {
  return inventoryItem
}
```

also the comment to me, the beginning of it causes some doubts "if adding a default lot item, but the existing inventoryItem...." 🤔 

let's do:
```
// If we search for an inventoryItem with default lot and no expiration date, and we find one with expirationDate, do not use it, and create a new one without the expiration date.

```

File: grails-app/domain/org/pih/warehouse/product/Product.groovy (hunk 1 above)
Comment (kchelstowski) [reply grails-app/domain/org/pih/warehouse/product/Product.groovy]: moreover - shouldn't this be called only if `inventoryItemsWithNullLot` size is > 1?

File: src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy [hunk 5]
```diff
@@ -91,7 +98,8 @@ class InventoryItemManager {
         }
 
         if (inventoryItems?.size() > 1) {
-            return inventoryItems.find { it.lotNumber == null } ?: inventoryItems.get(0)
+            // Prefer the item without expiration date.
+            return inventoryItems.find { it.expirationDate == null }
```
Comment (kchelstowski) [inline src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: why did you get rid of the fallback here? don't we break any existing behavior?

File: src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy [hunk 6]
```diff
@@ -123,7 +131,7 @@ class InventoryItemManager {
 
     private String sanitizeLotNumber(String lotNumber) {
         if (StringUtils.isEmpty(lotNumber)) {
-            return lotNumber
+            return null
```
Comment (kchelstowski) [inline src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: what does this change change?

File: grails-app/domain/org/pih/warehouse/product/Product.groovy (hunk 1 above)
Comment (SebastianLib) [reply grails-app/domain/org/pih/warehouse/product/Product.groovy]: If all of them have an expiration date we will have the fallback ?: inventoryItemsWithNullLot.get(0) and we just return the first one, so this is the same as the original behavior before this change. I think there's no way to decide which one is "better" in that case. In practice this is an extremely rare scenario, we only have one product in the database that actually has two default lot items both with expiry dates.

File: src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy (hunk 5 above)
Comment (SebastianLib) [reply src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: That's intentional. When all existing default lot items have an expiration date, we want to return null here so that getOrCreateInventoryItem proceeds to create a new item without an expiration date. Returning the fallback inventoryItems.get(0) (item with expiry) will be wrong and that is exactly the case we are trying to avoid

File: src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy (hunk 6 above)
Comment (SebastianLib) [reply src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: Without this change, if we already have an inventory item with an empty-string lot number and an expiration date, and we try to create a new one with an empty-string lot number but without an expiration date, we would get a unique constraint violation

Image

Comment (kchelstowski) [reply src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: I would like @ewaterman to comment on that. I understand the error, I just don't know if we don't break anything with that change to return null.

File: grails-app/utils/org/pih/warehouse/DateUtil.groovy [hunk 7]
```diff
@@ -51,6 +52,10 @@ class DateUtil {
         return ZoneId.systemDefault()
     }
 
+    static LocalDate toLocalDate(Date date) {
+        return date?.toInstant()?.atZone(getSystemZoneId())?.toLocalDate()
+    }
```
Comment (ewaterman) [inline grails-app/utils/org/pih/warehouse/DateUtil.groovy]: We already have `LocalDateParser.asLocalDate()`

File: src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy (hunk 6 above)
Comment (ewaterman) [reply src/main/groovy/org/pih/warehouse/inventory/InventoryItemManager.groovy]: yeah I feel uncomfortable with many of these changes. We should be striving for an InventoryItem to be unique on product + lot number and this feels like a step in the opposite direction.

InventoryItem has:
```
    static constraints = {
        product(nullable: false)
        lotNumber(nullable: true, unique: ['product'], maxSize: 255)
    }
```
The only reason why you're allowed to set multiple null lots is because Grails validation unique check excludes nulls, which in this case is a mistake. We should only allow one null (default) lot number per product.

Maybe I'm not understanding the issue, but to me it seems like a data issue. We should not have default lots with an expiry, or have two default lots, one with an expiry and one without.

So we should:
1) find everywhere in the code we're allowing that to happen and prevent it
2) fix the data so that we only have a single default lot on each product.

but that is not going to be a simple task.

Duplicate lots is the main reason why I wanted to rework the import feature. We have a ton of data issues (null vs "" vs " " lot products) that we need to fix.

so TLDR: I see this as a data issue, and we shouldn't be allowing multiple default lots to be created for a product because that just makes the problem worse

**Review by ewaterman — CHANGES_REQUESTED** [review]

File: grails-app/domain/org/pih/warehouse/product/Product.groovy (hunk 1 above)
Comment (jmiranda) [reply grails-app/domain/org/pih/warehouse/product/Product.groovy]: > we will have the fallback ?: inventoryItemsWithNullLot.get(0) 

Unfortunately, this might be too arbitrary to be viable. We need a dedicated "default" that is consistent. The default product lot **should** be the first InventoryItem created for this product, but even that I assume is too arbitrary.

**Review by jmiranda — COMMENTED** [review]
We're discussing the issue in Slack. This is a bigger problem than originally estimated so we're probably going to move this to the next release so we can have more discussions. 

This issue shares DNA with this ticket https://pihemr.atlassian.net/browse/OBPIH-7658 in case you want to take see a discussion of the problem as well as some possible solutions. If we work on OBPIH-7830, we need a full discovery that takes OBPIH-7658 (as well as some old OBS tickets) into account.

In the short-term, if we could fix the AvailableItemStatus issue, I think that might be good enough for now. The status resolution is too specific for what is actually happening here. 
```
        if (available && quantityAvailable < quantityOnHand) {
            return AvailableItemStatus.PICKED
        }
```

We should be using quantityAllocated and quantityPicked when resolving to allocated and/or picked status. I think both should be taken into account, but that might be a separate ticket.

And secondly, the fallback case at the end of the getStatus() method is wrong. ```
return AvailableItemStatus.PICKED
```
This should just resolve to Unavailable if there's not a better status. 
```
return AvailableItemStatus.NOT_AVAILABLE
```
Lastly, this should probably resolve to Out of Stock 
```
if (quantityOnHand <= 0) {
  return AvailableItemStatus.NOT_AVAILABLE
}
```

The only other thing worth exploring is whether product availability is out of whack because of the bad data (i.e. lotNumber is null, expirationDate is not null) and whether there's a short-term fix for that. If we're grouping by inventoryItem.lotNumber rather than inventoryItem.id then that's probably worth looking into to make sure the result set includes a row per inventoryItem.id rather than lotNumber (which aggregates with similar values). 

cc @ewaterman @awalkowiak Let me know if we should discuss either of these on a tech huddle tomorrow or just punt the whole thing to the next release.

File: src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy [hunk 8]
```diff
@@ -379,17 +379,17 @@ class AvailableItem {
         if (onHold) {
             return AvailableItemStatus.HOLD
         }
+        if (quantityOnHand <= 0) {
```
Comment (jmiranda) [inline src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy]: Can we make this `(quantityOnHand <= 0 || quantityAvailable <=0)`? Would that break anything? Otherwise we'll still see `Picked` in cases where the product availability has not been refreshed properly. 

This would not be necessary if we fixed the PICKED block below to handle quantity allocated and quantity picked, but since the AvailableItem doesn't include state for quantity picked and allocated we can punt this to another release.

**Review by jmiranda — CHANGES_REQUESTED** [review]
@SebastianLib Let me know your thoughts on this change? We probably want to add some tests to cover all of the cases we can think of including the edge cases (product availability not refreshed, test case related to this ticket).

cc @awalkowiak @ewaterman

File: src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy [hunk 9]
```diff
@@ -379,17 +379,17 @@ class AvailableItem {
...
         }
         if (available) {
             return AvailableItemStatus.AVAILABLE
         }
-        if (quantityOnHand <= 0) {
-            return AvailableItemStatus.NOT_AVAILABLE
-        }
 
-        return AvailableItemStatus.PICKED
+        return AvailableItemStatus.NOT_AVAILABLE
```
Comment (ewaterman) [inline src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy:392]: `StockMovementService.calculateAvailableItemsStatus` has some custom logic around PICKED and AVAILABLE (looks like it sets a pickedRequisitionNumbers field on the StockMovementItem). I'm not familiar with how that is being used. Can you take a look into what the impact will be of changing the default status here? From what I can see it is only used for display in the EditPickModal so likely this is fine, but it would be good to double check since we're at the end of release.

**Comment by SebastianLib** [comment]
> @SebastianLib Let me know your thoughts on this change? We probably want to add some tests to cover all of the cases we can think of including the edge cases (product availability not refreshed, test case related to this ticket).
> 
> cc @awalkowiak @ewaterman

@jmiranda I think this change is fine for now, at least the user won't be confused by seeing "Picked" status when there's nothing in pending outbound. I'll start writing the tests and try to cover as many edge cases as possible.

File: src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy (hunk 9 above)
Comment (SebastianLib) [reply src/main/groovy/org/pih/warehouse/api/StockMovementItem.groovy:392]: I tested this and for me everything looks fine. I'll also ask Kuba to pay attention to this during testing.

**Review by ewaterman — APPROVED** [review]
nice work on the tests!

**Review by alannadolny — APPROVED** [review]
